# MODELOS DE PREDICCION DE PEPTIDOS ANTIMICROBIANOS

#### - Edwin Villarraga
#### - Santiago Giraldo
#### - Mauricio
#### - Lesty
#### - Stephany

## <font color='red'>PARTE 1</font>

 ### <font color='blue'>PUNTO 1</font>

## Librerias

In [124]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso
from sklearn.feature_selection import SelectFromModel

## Cargar Datos

In [125]:
#Se cargan los datos negativos
dataN = pd.read_csv('./Datos/DatosNegativos2.csv')

In [126]:
#Se cargan los datos positivos
dataP = pd.read_csv('./Datos/DatosPositivos2.csv')

In [127]:
print("El archivo con datos Negativos tiene {} filas y {} columnas".format(dataN.shape[0],dataN.shape[1]))
print("El archivo con datos Positivos tiene {} filas y {} columnas".format(dataP.shape[0],dataP.shape[1]))

El archivo con datos Negativos tiene 5561 filas y 1761 columnas
El archivo con datos Positivos tiene 8322 filas y 1761 columnas


In [128]:
pd.options.display.max_columns = None
# Modificación de las opciones de pandas para mostrar todas las columnas del dataframe

In [129]:
# Verificación de campos
dataP.head(5)

Unnamed: 0                   sequence  length  molecular_weight  charge  \
0           0        CGGYSGGWKRKRSTSYRCG      19           2109.36   4.859   
1           1           GIMISLMKKLAAHIAK      16           1725.22   3.094   
2           2          FKAWRWAWRMKKLAAPS      17           2133.57   4.995   
3           3       WGWRDIVAGIRKVAAPVLST      20           2195.57   1.996   
4           4  DTHFPICIFCCGCCRQSNCGMCCKT      25           2774.34   0.555   

   charge_density  isoelectric_point     gravy  instability_index  \
0        0.002304          10.056702 -1.310526          73.031579   
1        0.001793          10.302063  0.887500          20.918750   
2        0.002341          12.024475 -0.594118          43.088235   
3        0.000909          10.835022  0.435000          17.010000   
4        0.000200           7.713074  0.392000          31.272000   

   aromaticity  aliphatic_index  boman_index  hydrophobic_ratio       A  \
0     0.157895         0.000000     3.122105           0.105263   0.000   
1     0.000000       140.625000    -0.685000           0.625000  18.750   
2     0.235294        46.470588     1.494706           0.411765  23.529   
3     0.100000       117.000000     0.561500           0.450000  15.000   
4     0.080000        31.200000     0.972000           0.520000   0.000   

        R    N    D       C    E    Q       G     H      I       L       K  \
0  15.789  0.0  0.0  10.526  0.0  0.0  26.316  0.00   0.00   0.000  10.526   
1   0.000  0.0  0.0   0.000  0.0  0.0   6.250  6.25  18.75  12.500  18.750   
2  11.765  0.0  0.0   0.000  0.0  0.0   0.000  0.00   0.00   5.882  17.647   
3  10.000  0.0  5.0   0.000  0.0  0.0  10.000  0.00  10.00   5.000   5.000   
4   4.000  4.0  4.0  32.000  0.0  4.0   8.000  4.00   8.00   0.000   4.000   

        M      F      P       S      T       W       Y     V    AA   AR   AN  \
0   0.000  0.000  0.000  15.789  5.263   5.263  10.526   0.0  0.00  0.0  0.0   
1  12.500  0.000  0.000   6.250  0.000   0.000   0.000   0.0  6.67  0.0  0.0   
2   5.882  5.882  5.882   5.882  0.000  17.647   0.000   0.0  6.25  0.0  0.0   
3   0.000  0.000  5.000   5.000  5.000  10.000   0.000  15.0  5.26  0.0  0.0   
4   4.000  8.000  4.000   4.000  8.000   0.000   0.000   0.0  0.00  0.0  0.0   

    AD   AC   AE   AQ    AG    AH   AI   AL    AK   AM   AF    AP   AS   AT  \
0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.0  0.00  0.0  0.0  0.00  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.00  6.67  0.0  0.0  6.67  0.0  0.0  0.00  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.0  0.00  0.0  0.0  6.25  0.0  0.0   
3  0.0  0.0  0.0  0.0  5.26  0.00  0.0  0.0  0.00  0.0  0.0  5.26  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.0  0.00  0.0  0.0  0.00  0.0  0.0   

     AW   AY   AV   RA   RR   RN    RD    RC   RE    RQ   RG   RH   RI   RL  \
0   0.0  0.0  0.0  0.0  0.0  0.0  0.00  5.56  0.0  0.00  0.0  0.0  0.0  0.0   
1   0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.00  0.0  0.0  0.0  0.0   
2  12.5  0.0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  0.00  0.0  0.0  0.0  0.0   
3   0.0  0.0  0.0  0.0  0.0  0.0  5.26  0.00  0.0  0.00  0.0  0.0  0.0  0.0   
4   0.0  0.0  0.0  0.0  0.0  0.0  0.00  0.00  0.0  4.17  0.0  0.0  0.0  0.0   

     RK    RM   RF   RP    RS   RT    RW   RY   RV   NA   NR   NN   ND    NC  \
0  5.56  0.00  0.0  0.0  5.56  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
1  0.00  0.00  0.0  0.0  0.00  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
2  0.00  6.25  0.0  0.0  0.00  0.0  6.25  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
3  5.26  0.00  0.0  0.0  0.00  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.00   
4  0.00  0.00  0.0  0.0  0.00  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  4.17   

    NE   NQ   NG   NH   NI   NL   NK   NM   NF   NP   NS   NT   NW   NY   NV  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 

In [130]:
# Verificación de campos
dataN.head(2)

Unnamed: 0 sequence  length  molecular_weight  charge  charge_density  \
0           0  MAATRNK       7            790.93   1.996        0.002524   
1           1  MKMDGLR       7            850.06   0.996        0.001172   

   isoelectric_point     gravy  instability_index  aromaticity  \
0          11.000305 -1.014286          77.657143          0.0   
1           8.497986 -0.671429          82.657143          0.0   

   aliphatic_index  boman_index  hydrophobic_ratio       A       R       N  \
0        28.571429     3.387143           0.428571  28.571  14.286  14.286   
1        55.714286     2.661429           0.428571   0.000  14.286   0.000   

        D    C    E    Q       G    H    I       L       K       M    F    P  \
0   0.000  0.0  0.0  0.0   0.000  0.0  0.0   0.000  14.286  14.286  0.0  0.0   
1  14.286  0.0  0.0  0.0  14.286  0.0  0.0  14.286  14.286  28.571  0.0  0.0   

     S       T    W    Y    V     AA   AR   AN   AD   AC   AE   AQ   AG   AH  \
0  0.0  14.286  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0   0.000  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    AI   AL   AK   AM   AF   AP   AS     AT   AW   AY   AV   RA   RR     RN  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0  0.0  16.67   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0  0.0   0.00   

    RD   RC   RE   RQ   RG   RH   RI   RL   RK   RM   RF   RP   RS   RT   RW  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    RY   RV   NA   NR   NN   ND   NC   NE   NQ   NG   NH   NI   NL     NK  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00   

    NM   NF   NP   NS   NT   NW   NY   NV   DA   DR   DN   DD   DC   DE   DQ  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

      DG   DH   DI   DL   DK   DM   DF   DP   DS   DT   DW   DY   DV   CA  \
0   0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  16.67  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    CR   CN   CD   CC   CE   CQ   CG   CH   CI   CL   CK   CM   CF   CP   CS  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    CT   CW   CY   CV   EA   ER   EN   ED   EC   EE   EQ   EG   EH   EI   EL  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    EK   EM   EF   EP   ES   ET   EW   EY   EV   QA   QR   QN   QD   QC   QE  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    QQ   QG   QH   QI   QL   QK   QM   QF   QP   QS   QT   QW   QY   QV   GA  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    GR   GN   GD   GC   GE   GQ   GG   GH   GI     GL   GK   GM   GF   GP  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.00  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  16.67  0.0  0.0  0.0  0.0   

    GS   GT   GW   GY   GV   HA   HR   HN   HD   HC   HE   HQ   HG   HH   HI  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    HL   HK   HM   HF   HP   HS   HT   HW   HY   HV   IA   IR   IN   ID   IC  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 

No se logra ver en las 1761 columnas la variable respuesta, entonces se adicionará la columna Respuesta=0 en el caso de los negativos y Respuesta=1 en el caso de los positivos.  Luego se unirán los datos en una nuevo Dataframe unificado

In [131]:
dataN['Respuesta']=0

In [132]:
dataP['Respuesta']=1

In [133]:
data = dataN.append(dataP) #se unen los datos de las respuestas positivas y negativas

In [145]:
len_id=data['sequence'].unique()

In [146]:
len_id

array(['MAATRNK', 'MKMDGLR', 'MNIKHYQ', ..., 'KKCFRWQWRMKKLGA',
       'GWAGWLLSPRGSRPSWGP', 'LRRLYKRVVKLWKRLFRQLRRPVR'], dtype=object)

In [149]:
len(data)

13883

In [150]:
len(data)-len(len_id)

81

Existen 81 datos de con sequence repetida

In [ ]:
#Se extrae una lista de las variables que contengan valores nulos
vars_with_na = [var for var in data.columns if data[var].isnull().sum()>1]
vars_with_na

Significa que no hay variables con valores nulos

## Análisis Descriptivo de los Datos

In [63]:
data.shape

(13883, 1762)

In [64]:
# lista de variables numericas
num_vars = [var for var in data.columns if data[var].dtypes != 'O']  #'O' significa python objects
len(num_vars)

1761

Todas las variables son numéricas excepto 1, que es la sequencia, que actua como identificador de la molécula

## Escalamiento

In [65]:
# Se excluye la variable Sequencia
data.drop(['sequence'], axis=1, inplace=True)

In [66]:
data.shape

(13883, 1761)

In [67]:
Xdata = data.drop(['Respuesta'], axis=1, inplace=False) #se crea un dataframe sin la variable respuesta
Ydata = data['Respuesta']
print(Xdata.shape)
print(Ydata.shape)

(13883, 1760)
(13883,)


In [68]:
train_vars = [var for var in Xdata.columns]
all_vars = [var for var in data.columns]

In [69]:
#ajuste del escalador
scaler = StandardScaler()  #se crea una instancia del escalador
scaler.fit(Xdata)

Xdata_std = pd.DataFrame(scaler.transform(Xdata), columns=train_vars) #  se ajusta el escalador a los datos de entrenamiento


In [70]:
type(Xdata_std)

pandas.core.frame.DataFrame

In [71]:
Xdata_std.head(2)

Unnamed: 0    length  molecular_weight    charge  charge_density  \
0   -1.621112 -1.567719         -1.647289 -0.019215        1.094644   
1   -1.620663 -1.567719         -1.575014 -0.353870        0.131734   

   isoelectric_point     gravy  instability_index  aromaticity  \
0           0.894593 -0.945354           0.870319    -1.102753   
1          -0.056603 -0.584815           0.984439    -1.102753   

   aliphatic_index  boman_index  hydrophobic_ratio         A         R  \
0        -1.266385      1.14450           0.091409  2.584165  0.602939   
1        -0.701275      0.73767           0.091409 -0.863028  0.602939   

          N         D        C         E         Q         G         H  \
0  2.317285 -0.570322 -0.51604 -0.579515 -0.577311 -0.921235 -0.467583   
1 -0.674234  2.339137 -0.51604 -0.579515 -0.577311  1.029598 -0.467583   

          I         L         K         M        F         P         S  \
0 -0.886105 -1.068507  0.403486  2.636298 -0.78091 -0.666612 -0.850156   
1 -0.886105  0.340929  0.403486  5.992251 -0.78091 -0.666612 -0.850156   

          T         W         Y        V        AA        AR        AN  \
0  1.821832 -0.448455 -0.590342 -0.87664  5.865496 -0.226733 -0.170002   
1 -0.751386 -0.448455 -0.590342 -0.87664 -0.296462 -0.226733 -0.170002   

         AD       AC        AE        AQ       AG        AH        AI  \
0 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   
1 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   

         AL        AK        AM        AF        AP        AS         AT  \
0 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  11.044019   
1 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  -0.209085   

         AW        AY       AV        RA        RR         RN        RD  \
0 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  13.069291 -0.164851   
1 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  -0.193227 -0.164851   

         RC        RE        RQ        RG        RH        RI        RL  \
0 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   
1 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   

         RK        RM        RF       RP       RS        RT        RW  \
0 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   
1 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   

         RY       RV       NA        NR        NN        ND        NC  \
0 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   
1 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   

         NE       NQ        NG        NH        NI        NL         NK  \
0 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  14.424998   
1 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  -0.185468   

         NM        NF        NP        NS        NT        NW        NY  \
0 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   
1 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   

         NV        DA        DR       DN        DD        DC        DE  \
0 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   
1 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   

         DQ         DG        DH        DI        DL        DK       DM  \
0 -0.099722  -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   
1 -0.099722  14.320288 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   

         DF        DP        DS        DT       DW        DY        DV  \
0 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   
1 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   

         CA        CR        CN        CD        CC        CE       CQ  \
0 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   
1 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   

         C

In [72]:
Xdata_std.head(2)

Unnamed: 0    length  molecular_weight    charge  charge_density  \
0   -1.621112 -1.567719         -1.647289 -0.019215        1.094644   
1   -1.620663 -1.567719         -1.575014 -0.353870        0.131734   

   isoelectric_point     gravy  instability_index  aromaticity  \
0           0.894593 -0.945354           0.870319    -1.102753   
1          -0.056603 -0.584815           0.984439    -1.102753   

   aliphatic_index  boman_index  hydrophobic_ratio         A         R  \
0        -1.266385      1.14450           0.091409  2.584165  0.602939   
1        -0.701275      0.73767           0.091409 -0.863028  0.602939   

          N         D        C         E         Q         G         H  \
0  2.317285 -0.570322 -0.51604 -0.579515 -0.577311 -0.921235 -0.467583   
1 -0.674234  2.339137 -0.51604 -0.579515 -0.577311  1.029598 -0.467583   

          I         L         K         M        F         P         S  \
0 -0.886105 -1.068507  0.403486  2.636298 -0.78091 -0.666612 -0.850156   
1 -0.886105  0.340929  0.403486  5.992251 -0.78091 -0.666612 -0.850156   

          T         W         Y        V        AA        AR        AN  \
0  1.821832 -0.448455 -0.590342 -0.87664  5.865496 -0.226733 -0.170002   
1 -0.751386 -0.448455 -0.590342 -0.87664 -0.296462 -0.226733 -0.170002   

         AD       AC        AE        AQ       AG        AH        AI  \
0 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   
1 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   

         AL        AK        AM        AF        AP        AS         AT  \
0 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  11.044019   
1 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  -0.209085   

         AW        AY       AV        RA        RR         RN        RD  \
0 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  13.069291 -0.164851   
1 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  -0.193227 -0.164851   

         RC        RE        RQ        RG        RH        RI        RL  \
0 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   
1 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   

         RK        RM        RF       RP       RS        RT        RW  \
0 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   
1 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   

         RY       RV       NA        NR        NN        ND        NC  \
0 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   
1 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   

         NE       NQ        NG        NH        NI        NL         NK  \
0 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  14.424998   
1 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  -0.185468   

         NM        NF        NP        NS        NT        NW        NY  \
0 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   
1 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   

         NV        DA        DR       DN        DD        DC        DE  \
0 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   
1 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   

         DQ         DG        DH        DI        DL        DK       DM  \
0 -0.099722  -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   
1 -0.099722  14.320288 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   

         DF        DP        DS        DT       DW        DY        DV  \
0 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   
1 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   

         CA        CR        CN        CD        CC        CE       CQ  \
0 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   
1 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   

         C

In [73]:
Xdata_std.tail(2)

Unnamed: 0    length  molecular_weight    charge  charge_density  \
13881    2.118256 -0.083772         -0.209532 -0.019215        0.019881   
13882    2.118705  0.725653          1.342546  2.992012        1.716255   

       isoelectric_point     gravy  instability_index  aromaticity  \
13881           1.274625 -0.527229           0.547323     0.537596   
13882           1.457403 -0.781359           2.059001     0.127509   

       aliphatic_index  boman_index  hydrophobic_ratio         A         R  \
13881        -0.843379    -0.204926          -1.670353 -0.192676  0.296353   
13882         0.585092     1.573553          -0.268951 -0.863028  2.442163   

              N         D        C         E         Q         G         H  \
13881 -0.674234 -0.570322 -0.51604 -0.579515 -0.577311  2.113303 -0.467583   
13882 -0.674234 -0.570322 -0.51604 -0.579515  0.327886 -0.921235 -0.467583   

              I         L         K         M         F         P         S  \
13881 -0.886105  0.027688 -0.874347 -0.719891 -0.780910  1.838820  1.766510   
13882 -0.886105  0.986847  0.243734 -0.719891 -0.157235 -0.040216 -0.850156   

              T         W         Y        V        AA        AR        AN  \
13881 -0.751386  1.758806 -0.590342 -0.87664 -0.296462 -0.226733 -0.170002   
13882 -0.751386  0.103393  0.355567  1.00513 -0.296462 -0.226733 -0.170002   

             AD       AC        AE        AQ        AG        AH        AI  \
13881 -0.168532 -0.14897 -0.179553 -0.152015  2.781368 -0.145561 -0.255284   
13882 -0.168532 -0.14897 -0.179553 -0.152015 -0.288130 -0.145561 -0.255284   

             AL        AK        AM        AF        AP        AS        AT  \
13881 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932 -0.209085   
13882 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932 -0.209085   

             AW        AY       AV        RA        RR        RN        RD  \
13881 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051 -0.193227 -0.164851   
13882 -0.138379 -0.144547 -0.26498 -0.205787  2.107522 -0.193227 -0.164851   

             RC        RE        RQ        RG        RH        RI        RL  \
13881 -0.196828 -0.157531 -0.164182  3.140833 -0.133507 -0.237727 -0.280231   
13882 -0.196828 -0.157531  3.941452 -0.244849 -0.133507 -0.237727  3.051372   

             RK        RM        RF        RP       RS        RT        RW  \
13881 -0.271849 -0.116979 -0.224233  2.831458 -0.20059 -0.190032 -0.192644   
13882 -0.271849 -0.116979 -0.224233  2.048374 -0.20059 -0.190032 -0.192644   

             RY        RV       NA        NR        NN        ND        NC  \
13881 -0.157728 -0.220710 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   
13882 -0.157728  2.099535 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   

             NE       NQ        NG        NH        NI        NL        NK  \
13881 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701 -0.185468   
13882 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701 -0.185468   

             NM        NF        NP        NS        NT        NW        NY  \
13881 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   
13882 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   

             NV        DA        DR       DN        DD        DC        DE  \
13881 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   
13882 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   

             DQ        DG        DH        DI        DL        DK       DM  \
13881 -0.099722 -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   
13882 -0.099722 -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   

             DF        DP        DS        DT       DW        DY        DV  \
13881 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   
13882 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   

             CA        CR        

In [74]:
type(Xdata_std)

pandas.core.frame.DataFrame

 ### <font color='blue'>PUNTO 2</font>

### Dividir los datos en train y test

In [75]:
type(Ydata)

pandas.core.series.Series

In [76]:
Ydata.shape

(13883,)

In [77]:
Xdata_std.shape

(13883, 1760)

In [78]:
Xdata_std.head(2)

Unnamed: 0    length  molecular_weight    charge  charge_density  \
0   -1.621112 -1.567719         -1.647289 -0.019215        1.094644   
1   -1.620663 -1.567719         -1.575014 -0.353870        0.131734   

   isoelectric_point     gravy  instability_index  aromaticity  \
0           0.894593 -0.945354           0.870319    -1.102753   
1          -0.056603 -0.584815           0.984439    -1.102753   

   aliphatic_index  boman_index  hydrophobic_ratio         A         R  \
0        -1.266385      1.14450           0.091409  2.584165  0.602939   
1        -0.701275      0.73767           0.091409 -0.863028  0.602939   

          N         D        C         E         Q         G         H  \
0  2.317285 -0.570322 -0.51604 -0.579515 -0.577311 -0.921235 -0.467583   
1 -0.674234  2.339137 -0.51604 -0.579515 -0.577311  1.029598 -0.467583   

          I         L         K         M        F         P         S  \
0 -0.886105 -1.068507  0.403486  2.636298 -0.78091 -0.666612 -0.850156   
1 -0.886105  0.340929  0.403486  5.992251 -0.78091 -0.666612 -0.850156   

          T         W         Y        V        AA        AR        AN  \
0  1.821832 -0.448455 -0.590342 -0.87664  5.865496 -0.226733 -0.170002   
1 -0.751386 -0.448455 -0.590342 -0.87664 -0.296462 -0.226733 -0.170002   

         AD       AC        AE        AQ       AG        AH        AI  \
0 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   
1 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   

         AL        AK        AM        AF        AP        AS         AT  \
0 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  11.044019   
1 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  -0.209085   

         AW        AY       AV        RA        RR         RN        RD  \
0 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  13.069291 -0.164851   
1 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  -0.193227 -0.164851   

         RC        RE        RQ        RG        RH        RI        RL  \
0 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   
1 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   

         RK        RM        RF       RP       RS        RT        RW  \
0 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   
1 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   

         RY       RV       NA        NR        NN        ND        NC  \
0 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   
1 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   

         NE       NQ        NG        NH        NI        NL         NK  \
0 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  14.424998   
1 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  -0.185468   

         NM        NF        NP        NS        NT        NW        NY  \
0 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   
1 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   

         NV        DA        DR       DN        DD        DC        DE  \
0 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   
1 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   

         DQ         DG        DH        DI        DL        DK       DM  \
0 -0.099722  -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   
1 -0.099722  14.320288 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   

         DF        DP        DS        DT       DW        DY        DV  \
0 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   
1 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   

         CA        CR        CN        CD        CC        CE       CQ  \
0 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   
1 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   

         C

In [79]:
Xdata_std.reset_index(drop=True, inplace=True)
Ydata.to_frame()
Ydata.reset_index(drop=True, inplace=True)

In [80]:
Data = pd.concat([Xdata_std,Ydata], axis=1)

In [81]:
Data.head(2)

Unnamed: 0    length  molecular_weight    charge  charge_density  \
0   -1.621112 -1.567719         -1.647289 -0.019215        1.094644   
1   -1.620663 -1.567719         -1.575014 -0.353870        0.131734   

   isoelectric_point     gravy  instability_index  aromaticity  \
0           0.894593 -0.945354           0.870319    -1.102753   
1          -0.056603 -0.584815           0.984439    -1.102753   

   aliphatic_index  boman_index  hydrophobic_ratio         A         R  \
0        -1.266385      1.14450           0.091409  2.584165  0.602939   
1        -0.701275      0.73767           0.091409 -0.863028  0.602939   

          N         D        C         E         Q         G         H  \
0  2.317285 -0.570322 -0.51604 -0.579515 -0.577311 -0.921235 -0.467583   
1 -0.674234  2.339137 -0.51604 -0.579515 -0.577311  1.029598 -0.467583   

          I         L         K         M        F         P         S  \
0 -0.886105 -1.068507  0.403486  2.636298 -0.78091 -0.666612 -0.850156   
1 -0.886105  0.340929  0.403486  5.992251 -0.78091 -0.666612 -0.850156   

          T         W         Y        V        AA        AR        AN  \
0  1.821832 -0.448455 -0.590342 -0.87664  5.865496 -0.226733 -0.170002   
1 -0.751386 -0.448455 -0.590342 -0.87664 -0.296462 -0.226733 -0.170002   

         AD       AC        AE        AQ       AG        AH        AI  \
0 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   
1 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   

         AL        AK        AM        AF        AP        AS         AT  \
0 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  11.044019   
1 -0.330038 -0.279003 -0.124839 -0.205495 -0.184651 -0.254932  -0.209085   

         AW        AY       AV        RA        RR         RN        RD  \
0 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  13.069291 -0.164851   
1 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051  -0.193227 -0.164851   

         RC        RE        RQ        RG        RH        RI        RL  \
0 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   
1 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   

         RK        RM        RF       RP       RS        RT        RW  \
0 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   
1 -0.271849 -0.116979 -0.224233 -0.17804 -0.20059 -0.190032 -0.192644   

         RY       RV       NA        NR        NN        ND        NC  \
0 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   
1 -0.157728 -0.22071 -0.18812 -0.179051 -0.147625 -0.125675 -0.122088   

         NE       NQ        NG        NH        NI        NL         NK  \
0 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  14.424998   
1 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701  -0.185468   

         NM        NF        NP        NS        NT        NW        NY  \
0 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   
1 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   

         NV        DA        DR       DN        DD        DC        DE  \
0 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   
1 -0.169148 -0.176683 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   

         DQ         DG        DH        DI        DL        DK       DM  \
0 -0.099722  -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   
1 -0.099722  14.320288 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   

         DF        DP        DS        DT       DW        DY        DV  \
0 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   
1 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   

         CA        CR        CN        CD        CC        CE       CQ  \
0 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   
1 -0.158739 -0.184789 -0.130264 -0.105332 -0.110221 -0.100515 -0.10669   

         C

In [82]:
Data.drop(Data.columns[0], axis=1, inplace = True)

In [83]:
# Separación de los datos en train y test
# random_state para reproducibilidad

X_train, X_test, y_train, y_test = train_test_split(Data, Data.Respuesta,
                                                    test_size=0.2,
                                                    random_state=0)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((11106, 1760), (2777, 1760), (11106,), (2777,))

In [84]:
#eliminar la variable respuesta de X_train y X_test
X_train.drop(['Respuesta'], axis=1, inplace=True)
X_test.drop(['Respuesta',], axis=1, inplace=True)


C:\Users\Usuario\Anaconda3\envs\patrones\lib\site-packages\pandas\core\frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


 ### <font color='blue'>PUNTOS 3 y 4</font>

## MODELO 1: Regresión Logística

In [85]:
from sklearn.model_selection import KFold, cross_val_score
import warnings
warnings.filterwarnings('ignore')

In [86]:
X_train.head(2)

length  molecular_weight    charge  charge_density  \
3634   1.669983          1.388688 -0.688525       -0.703666   
12130  0.455845          0.426859 -0.221682       -0.304565   

       isoelectric_point     gravy  instability_index  aromaticity  \
3634           -1.002694  0.273887          -0.545338    -0.467779   
12130           0.048730  0.197717           0.247890     0.239351   

       aliphatic_index  boman_index  hydrophobic_ratio         A         R  \
3634          0.308054    -0.241454           0.246403  1.472222 -0.153531   
12130        -0.195651    -0.409287           0.266129  0.782204 -0.337675   

              N         D        C         E         Q         G         H  \
3634   0.001297  1.400477 -0.51604 -0.579515 -0.577311 -0.040178 -0.467583   
12130 -0.674234  0.355304 -0.51604 -0.579515 -0.577311  0.940837  3.109368   

              I         L         K         M        F         P         S  \
3634   0.560429 -0.113788 -0.585792  0.037989 -0.78091  0.788063 -0.343685   
12130 -0.206711 -0.171602 -0.467812  0.347859  0.57974 -0.666612 -0.136606   

              T         W         Y         V        AA        AR        AN  \
3634   0.410759 -0.448455  0.874263 -0.390993  0.934451 -0.226733 -0.170002   
12130 -0.751386  0.153453 -0.590342 -0.192429  1.463042 -0.226733 -0.170002   

             AD       AC        AE        AQ       AG        AH        AI  \
3634  -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561  1.656642   
12130 -0.168532 -0.14897 -0.179553 -0.152015 -0.28813 -0.145561 -0.255284   

             AL        AK        AM        AF        AP        AS        AT  \
3634   0.893334 -0.279003  3.817801 -0.205495 -0.184651 -0.254932  2.038836   
12130  1.418686  1.545184 -0.124839 -0.205495 -0.184651 -0.254932 -0.209085   

             AW        AY       AV        RA        RR        RN        RD  \
3634  -0.138379 -0.144547 -0.26498 -0.205787 -0.314051 -0.193227 -0.164851   
12130 -0.138379 -0.144547 -0.26498 -0.205787 -0.314051 -0.193227 -0.164851   

             RC        RE        RQ        RG        RH        RI        RL  \
3634  -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   
12130 -0.196828 -0.157531 -0.164182 -0.244849 -0.133507 -0.237727 -0.280231   

             RK        RM        RF        RP       RS        RT        RW  \
3634  -0.271849 -0.116979 -0.224233  3.235795 -0.20059 -0.190032 -0.192644   
12130 -0.271849 -0.116979 -0.224233 -0.178040 -0.20059 -0.190032 -0.192644   

             RY        RV        NA        NR        NN        ND        NC  \
3634  -0.157728 -0.220710  2.411333 -0.179051 -0.147625 -0.125675 -0.122088   
12130 -0.157728  2.318225 -0.188120 -0.179051 -0.147625 -0.125675 -0.122088   

             NE       NQ        NG        NH        NI        NL        NK  \
3634  -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701 -0.185468   
12130 -0.128434 -0.13149 -0.170668 -0.117616 -0.169302 -0.216701 -0.185468   

             NM        NF        NP        NS        NT        NW        NY  \
3634  -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   
12130 -0.100742 -0.159483 -0.139744 -0.176047 -0.156665 -0.119402 -0.124696   

             NV        DA        DR       DN        DD        DC        DE  \
3634  -0.169148  2.747450 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   
12130 -0.169148  4.003159 -0.147897 -0.11747 -0.111116 -0.092078 -0.129508   

             DQ        DG        DH        DI        DL        DK       DM  \
3634  -0.099722 -0.155138 -0.085028  2.950102 -0.201946 -0.175331 -0.08996   
12130 -0.099722 -0.155138 -0.085028 -0.167015 -0.201946 -0.175331 -0.08996   

             DF        DP        DS        DT       DW        DY        DV  \
3634  -0.155508  2.779367 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   
12130 -0.155508 -0.137355 -0.152566 -0.157472 -0.09771 -0.116464 -0.169329   

             CA        CR        CN        CD        CC        CE      

In [87]:
kfold = KFold(n_splits=10, random_state=0)
model_kfold = LogisticRegression()
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold,n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of  10 | elapsed:  1.1min remaining:   44.7s


Accuracy: 88.00%


[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:  1.4min finished


## Modelo 2 - Random Forest

In [88]:
from sklearn.ensemble import RandomForestClassifier

In [89]:
kfold = KFold(n_splits=10, random_state=0)
model_kfold = RandomForestClassifier(random_state=0)
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold, n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of  10 | elapsed:    3.8s remaining:    2.5s
[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:    5.8s finished


Accuracy: 89.28%


## Modelo 3 - Máquina de Vectores de Soporte

In [90]:
from sklearn import svm

In [91]:
kfold = KFold(n_splits=10, random_state=0)
model_kfold = svm.SVC(kernel='linear', max_iter=10000)
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold, n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of  10 | elapsed:  2.2min remaining:  1.4min


Accuracy: 83.75%


[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:  3.3min finished


## Modelo 4 - Adaboost

In [92]:
from sklearn.ensemble import AdaBoostClassifier

In [93]:
kfold = KFold(n_splits=10, random_state=0)
model_kfold = AdaBoostClassifier(n_estimators=100, random_state=0)
results_kfold = cross_val_score(model_kfold, X_train, y_train, cv=kfold, n_jobs=-1, verbose=1)
print("Accuracy: %.2f%%" % (results_kfold.mean()*100.0)) 

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of  10 | elapsed:  2.3min remaining:  1.5min


Accuracy: 89.13%


[Parallel(n_jobs=-1)]: Done  10 out of  10 | elapsed:  3.7min finished


 ### <font color='blue'>PUNTO 5</font>

Valiadación de los modelos con los datos de prueba

## MODELO 1: Regresión Logística

In [94]:
lr = LogisticRegression(solver='liblinear')
lr.fit(X_train, y_train) #con todas las variables
y_pred = lr.predict(X_test)
print("{0:.1%} precisión del modelo".format(accuracy_score(y_test, y_pred))) 

88.2% precisión del modelo


## MODELO 2 - Random Forest

In [95]:
clf = RandomForestClassifier(random_state=0)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 

89.2% accuracy on test set.


## MODELO 3 - Máquina de Vectores de Soporte

In [96]:
clf = svm.SVC(kernel='linear', max_iter=10000)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 

82.9% accuracy on test set.


## MODELO 4 - Adaboost

In [97]:
clf = AdaBoostClassifier(n_estimators=100, random_state=0)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 

89.0% accuracy on test set.


## <font color='red'>PARTE 2</font>

 ### <font color='blue'>PUNTO 1</font>

## Selección y Reducción de Variables

Se usará una regresión lasso,, con una semilla = 0, y la función selectFromModel de sklearn para seleccionar las variables más predictivas.  Entre mayor alpha más reducción de variables

In [98]:
sel_ = SelectFromModel(Lasso(alpha=0.01, random_state=0)) # remember to set the seed, the random state in this function
sel_.fit(X_train, y_train)

SelectFromModel(estimator=Lasso(alpha=0.01, copy_X=True, fit_intercept=True,
                                max_iter=1000, normalize=False, positive=False,
                                precompute=False, random_state=0,
                                selection='cyclic', tol=0.0001,
                                warm_start=False),
                max_features=None, norm_order=1, prefit=False, threshold=None)

In [99]:
sel_.get_support()  #para visualizar las varaibles seleccionas

array([False, False,  True, ..., False, False, False])

In [100]:
sum(sel_.get_support())  #se seleccionaron 15 variables significativas según el modelo RL Lasso

171

In [101]:
selected_feat = X_train.columns[(sel_.get_support())]

In [102]:
selected_feat  #caracteristicas seleccionadas

Index(['charge', 'charge_density', 'isoelectric_point', 'C', 'K', 'M', 'W',
       'Y', 'AM', 'AF',
       ...
       'embed_1_73', 'embed_1_93', 'embed_1_94', 'embed_1_99', 'embed_2_32',
       'embed_2_50', 'embed_2_51', 'embed_2_57', 'embed_2_73', 'embed_2_94'],
      dtype='object', length=171)

In [103]:
# algunas estadisticas
print('total variables: {}'.format((X_train.shape[1])))
print('varaibles seleccionadas: {}'.format(len(selected_feat)))
print('Variables excluidas por RL Lasso: {}'.format(
    np.sum(sel_.estimator_.coef_ == 0)))

total variables: 1759
varaibles seleccionadas: 171
Variables excluidas por RL Lasso: 1585


In [104]:
selected_feats = X_train.columns[(sel_.estimator_.coef_ != 0).ravel().tolist()]
selected_feats  #otra forma de ver las caracteriticas seleccionadas

Index(['charge', 'charge_density', 'isoelectric_point', 'C', 'K', 'M', 'W',
       'Y', 'AM', 'AF',
       ...
       'embed_1_73', 'embed_1_93', 'embed_1_94', 'embed_1_99', 'embed_2_32',
       'embed_2_50', 'embed_2_51', 'embed_2_57', 'embed_2_73', 'embed_2_94'],
      dtype='object', length=174)

In [105]:
# nse guarda la lista de las caracteristicas seleccionadas
pd.Series(selected_feats).to_csv('./Datos/selected_features.csv', index=False)

In [106]:
Xtrain = X_train[selected_feats]
Xtest = X_test[selected_feats]
#Se guarda la base de datos reducida co las variables seleccionadas
Xtrain.to_csv('./Datos/xtrain.csv', index=False)
Xtest.to_csv('./Datos/xtest.csv', index=False)

 ### <font color='blue'>PUNTO 2</font>

## MODELO 1 - Regresión Logística

In [107]:
lr = LogisticRegression(solver='liblinear')
lr.fit(Xtrain, y_train) #con todas las variables
y_pred = lr.predict(Xtest)
print("{0:.1%} precisión del modelo".format(accuracy_score(y_test, y_pred))) 

90.8% precisión del modelo


## MODELO 2 - Random Forest

In [108]:
clf = RandomForestClassifier(random_state=0)
clf.fit(Xtrain, y_train)
y_pred = clf.predict(Xtest)
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc))  

90.0% accuracy on test set.


## MODELO 4 - Adaboost

In [109]:
clf = AdaBoostClassifier(n_estimators=100, random_state=0)
clf.fit(Xtrain, y_train)
y_pred = clf.predict(Xtest)
# Calculates the accuracy on the test set and prints coefficients
acc = accuracy_score(y_test, y_pred)
print("{0:.1%} accuracy on test set.".format(acc)) 

90.0% accuracy on test set.
